File: Tree/94.Binary Tree Maximum Path Sum/Solution.java
/**
 * Definition of TreeNode:
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left, right;
 *     public TreeNode(int val) {
 *         this.val = val;
 *         this.left = this.right = null;
 *     }
 * }
 */


public class Solution {
    /**
     * @param root: The root of binary tree.
     * @return: An integer
     */
    private int max = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        // write your code here
        helper(root);
        return max;
    }
    
    private int helper(TreeNode root) {
        if (root == null) {
            return 0;
        }
        
        int left = helper(root.left);
        int right = helper(root.right);
        
        int curr_max = Math.max(Math.max(left + root.val, right + root.val), root.val);
        max = Math.max(max, Math.max(curr_max, root.val + left + right));
        return curr_max;
    }
}

File: Tree/94.Binary Tree Maximum Path Sum/Solution.py
import sys

"""
Definition of TreeNode:
class TreeNode:
    def __init__(self, val):
        self.val = val
        self.left, self.right = None, None
"""

class Solution:
    """
    @param root: The root of binary tree.
    @return: An integer
    """
    def maxPathSum(self, root):
        # write your code here
        self.maxSum = -sys.maxsize
        self._find_max_path_sum(root)
        return self.maxSum

    def _find_max_path_sum(self, root):
        if root is None:
            return 0 

        leftSum = self._find_max_path_sum(root.left)
        rightSum = self._find_max_path_sum(root.right)

        self.maxSum = max(self.maxSum, root.val + leftSum + rightSum, leftSum + root.val, rightSum + root.val, root.val)
        return max(root.val + leftSum, root.val + rightSum)
